Modules/grpmodule.c

```diff
@@ -119,16 +120,51 @@ grp_getgrgid_impl(PyObject *module, PyObject *id)
         }
         Py_DECREF(py_int_id);
     }
+#ifdef HAVE_GETGRGID_R
+    Py_BEGIN_ALLOW_THREADS
+    int status, bufsize = NSS_BUFLEN_GROUP;
+    struct group *grpbuf = NULL;
+
+    p = malloc(sizeof(struct group));
+    buf = malloc(bufsize);
+
+    do {
+        status = getgrgid_r(gid, p, buf, bufsize, &grpbuf);
+        if(grpbuf == NULL && status == ERANGE) {
```

Comment:
This condition is doubled in the `while` below. It would be better to write:
```c
if (grpbuf != NULL || status != ERANGE) {
    break;
}
```
and use `while (1)`.